### DISCUSSION TITLE
Reorganization of Parameters to Stay Below 25

### DISCUSSION THREAD

**Discussion by chrismuzyn** [Kilvoctu/aiyabot#224] [General]
Hi, 
This is pertaining to the 25 option limit that discord enforces.  I'm working on adding controlnet options for aiyabot.  There are a lot of options that can be used for the [controlnet extension](https://github.com/Mikubill/sd-webui-controlnet/wiki/API).

What is the strategy for adding new command options to aiyabot?  Should I make a new slashcommand for controlnet?  The problem with this route is that this new controlnet slashcommand *could* take all of the options for /draw in addition to the new controlnet options, so I would be left picking and choosing which options to include to stay under 25.  

Should we maybe have per feature settings?  /draw-settings, /controlnet-settings?  But it doesn't feel great to have to change channel settings every time you want to draw something, so maybe settings should be user specific?  But it would be a lot to start keeping track of every user's settings.

Does anyone have any ideas to solve this problem? Is there a way we can use forms or buttons maybe?  I am down to do work on a suggested solution.

**Comment by solareon** [comment]
I would say use a form for the controlnet input but I think that using controlnet via a discord bot might be a little more than what was intended. I believe there is some example form code when you do the edit for a redraw. Best of luck to ya

**Comment by wizz13150** [comment]
Yep, the good question is what CN options you want to use and the ones you can do without.

You can use a new command, to have 23-24 'presets' of CN units (as payloads). (1-2 choices for image acquisition, URL is better to use)
Or a sub-command for /Draw (didn't test this yet).
Or simply some options (until the 25th /draw option) to use some presets like the 'poseref' using the openpose model, and then pack these presets in a single option, but for other models, with decent attached/hard coded settings.
Using a form is another choice for a better UI experience, as mentionned.

But still, the first move is to choose "the ones you can do without" and then you may choose the best/easiest way to implement it.
(The generated pictures below are using the model [ZavyChromaXL](https://civitai.com/models/119229?modelVersionId=169740) and my current [version](https://github.com/wizz13150/aiyabot/tree/DiscordFullBot) of the bot ) 

# Poseref

Here is an implementation of the Poseref option for /Draw:
https://github.com/search?q=repo%3Awizz13150/aiyabot%20poseref&type=code

Source for my version (advo-kat fork):
https://github.com/advo-kat/aiyabot/commit/d1a25194c34538e05ff17a9fc01daba7217afcc0

Screen :
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/bf202abe-dc20-4eee-9a8f-7829aad64c41)

My CN Payload (not exhaustive code):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/f807cd08-2a4a-4898-a7ea-ad6d7a9901a6)

Source (poor quality, not the best source lol):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/b7604b67-c8cd-4f28-a58d-47a9dfdca0cc)

Result, Cyberpunk Version (Poseref for openpose):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/c63443d9-b964-4119-8af3-441ad11b31f9)


# ADetailer
You also may like the ADetailer option for /Draw (need to install 'ADetailer' extension in the webui):
https://github.com/search?q=repo%3Awizz13150%2Faiyabot+adetailer&type=code 

Screen:
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/1db5931c-d83f-48a8-ab80-dca9776827fa)

My ADetailer payload (not exhaustive code):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/fdb168f8-e02c-444d-845a-1cd84c115dfe)

Source for my version (also advo-kat fork):
https://github.com/advo-kat/aiyabot/commit/72595db479a6f8aa603d5e06b94dcabefdcf7968


Without ADetailer option (Face only):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/7b5761e4-5b93-4920-a7ef-213a65ac5ae1)

With ADetailer option (Face only):
![image](https://github.com/Kilvoctu/aiyabot/assets/22177081/87aebf54-d7c5-47a5-bd84-040f49991403)


My Civitai gallery, 100% of the images are from the Discord Bot:
https://civitai.com/user/_Wizz_/images

Cheers ! 🥂
